### ISSUE TITLE
Can we set an input to `None` such that the rule can never be used to avoid conditions that include/exclude rules from scope

### ISSUE THREAD

**Issue by nh13** [snakemake/snakemake#2526]
I would like Snakemake to treat `None` input values as though they are paths that will never exist.

For example, the `foo` rule below will only be used to generate the output if the variable `input_txt` is not `None`.  If `input_txt` is `None`, perhaps there is an alternate composition of rules that could produce "output.txt".  In this case, the `bar` rule can produce "output.txt" if the "input2.txt" is present.  The solution is to only include the `foo` rule in scope when `input_txt is not None`.  But I want to avoid conditional inclusion of rules as a pattern (use Snakemake's inference engine to figure out which rules to use!).

This is a contrived example, so I could make it work without an if/else, so is for illustrative purposes. 

```snakemake
from typing import Optional
input_txt: Optional[None] = config.get("input_txt", None)

rule all:
  input:
    txt = "output.txt"

rule foo:
  input:
    txt = input_txt
  output:
    txt = "output.txt"
  shell: 
    """cp -v {input.txt} {output.txt}"""

rule bar:
  input:
    txt = "input2.txt"
  output:
    txt = "output.txt"
  shell: 
    """cp -v {input.txt} {output.txt}"""
```

**Comment by AroneyS** [comment]
Can't you just use an ifelse on the output instead?
e.g. `output: "output.txt" if input_txt else []`

**Comment by SichongP** [comment]
> ```
> from typing import Optional
> input_txt: Optional[None] = config.get("input_txt", None)
> 
> rule all:
>   input:
>     txt = "output.txt"
> 
> rule foo:
>   input:
>     txt = input_txt
>   output:
>     txt = "output.txt"
>   shell: 
>     """cp -v {input.txt} {output.txt}"""
> 
> rule bar:
>   input:
>     txt = "input2.txt"
>   output:
>     txt = "output.txt"
>   shell: 
>     """cp -v {input.txt} {output.txt}"""
> ```

What if `input_txt` is not None? Wouldn't this lead to an AmbiguousRuleException seeing now you have two rules that make the same output file?

Using if-else in `output` would still have the same problem when `input_txt` isn't None.

One option is to use an input function on `rule all` with `foo` and `bar` producing different outputs:

```
rule all:
  input:
    txt = "output1.txt" if "input_txt" in config else "output2.txt"

rule foo:
  input:
    txt = config.get("input_txt")
  output:
    txt = "output1.txt"
  shell: 
    """cp -v {input.txt} {output.txt}"""

rule bar:
  input:
    txt = "input2.txt"
  output:
    txt = "output2.txt"
  shell: 
    """cp -v {input.txt} {output.txt}"""
```

**Comment by AroneyS** [comment]
The AmbiguousRuleException can be fixed by setting rule order at the top. Rule `foo` will be run instead of `bar` if both produce the same output. And then if input.txt is None, then they don't clash and the ruleorder is ignored.
e.g.

```python
ruleorder: foo > bar
```

**Comment by SichongP** [comment]
> The AmbiguousRuleException can be fixed by setting rule order at the top. Rule `foo` will be run instead of `bar` if both produce the same output. And then if input.txt is None, then they don't clash and the ruleorder is ignored. e.g.
> 
> ```python
> ruleorder: foo > bar
> ```

Yes that works too. Although I personally avoid `ruleorder` because imo they make snakefiles harder to follow. But that's just a preference.

**Comment by nh13** [comment]
Bump
